tests/ui/pattern/usefulness/impl-trait.rs

```diff
@@ -0,0 +1,120 @@
+#![feature(never_type)]
+#![feature(exhaustive_patterns)]
+#![feature(type_alias_impl_trait)]
+#![feature(non_exhaustive_omitted_patterns_lint)]
+#![deny(unreachable_patterns)]
+// Test that the lint traversal handles opaques correctly
+#![deny(non_exhaustive_omitted_patterns)]
+
+fn main() {}
+
+#[derive(Copy, Clone)]
+enum Void {}
+
+fn return_never_rpit(x: Void) -> impl Copy {
+    if false {
+        match return_never_rpit(x) {}
+    }
+    x
+}
+fn friend_of_return_never_rpit(x: Void) {
+    match return_never_rpit(x) {}
+    //~^ERROR non-empty
+}
+
+type T = impl Copy;
+fn return_never_tait(x: Void) -> T {
+    if false {
+        match return_never_tait(x) {}
+    }
+    x
+}
+fn friend_of_return_never_tait(x: Void) {
+    match return_never_tait(x) {}
+    //~^ERROR non-empty
+}
+
+fn option_never(x: Void) -> Option<impl Copy> {
+    if true {
+        match option_never(x) {
+            None => {}
+            Some(_) => {} //~ERROR unreachable
+        }
+        match option_never(x) {
+            None => {}
+            _ => {} //~ERROR unreachable
+        }
+        match option_never(x) {
+            None => {}
+        }
+    }
+    Some(x)
+}
+
+fn option_never2(x: Void) -> impl Copy {
+    if true {
+        match option_never2(x) {
+            None => {}
+            Some(_) => {} //~ERROR unreachable
+        }
+        match option_never2(x) {
+            None => {}
+            _ => {} //~ERROR unreachable
+        }
+        match option_never2(x) {
+            None => {}
+        }
+    }
+    Some(x)
+}
+
+fn inner_never(x: Void) {
+    type T = impl Copy;
+    let y: T = x;
+    match y {}
+}
+
+// This one caused ICE https://github.com/rust-lang/rust/issues/117100.
+fn inner_tuple() {
+    type T = impl Copy;
+    let foo: T = Some((1u32, 2u32));
+    match foo {
+        _ => {}
+        Some((a, b)) => {} //~ERROR unreachable
+    }
+}
+
+type U = impl Copy;
+fn unify_never(x: Void, u: U) -> U {
+    if true { match u {} } else { x }
+}
+
+type V = impl Copy;
+fn infer_in_match(x: Option<V>) {
+    match x {
+        None => {}
+        Some((a, b)) => {}
+        Some((mut x, mut y)) => {
+            //~^ERROR unreachable
+            x = 42;
+            y = "foo";
+        }
+    }
+}
+
+type W = impl Copy;
+#[derive(Copy, Clone)]
+struct Rec<'a> {
+    n: u32,
+    w: Option<&'a W>,
+}
```

Comment:
Oh good point! Typeck doesn't let me however, I'm not entirely sure why

Reply:
Actually what issue do you mean? I added a commit where I detect opaque cycles when revealing, so I won't crash there. Moreover it seems such a type is rejected by typeck anyway, even if the error is cryptic.

If you mean the ICEs because of unrevealed opaques in exhaustiveness, my conclusion is that `reveal_opaque_ty` simply cannot reveal all the things, so I need to keep revealing opaques at each depth.

Reply:
Oh wait, so this doesn't detect uninhabitedness everywhere actually:
```rust
type X = impl Copy;
struct SecretelyVoid(X);
fn nested_empty_opaque(x: Void) -> X {
    if true {
        let opaque_void = nested_empty_opaque(x);
        let secretely_void = SecretelyVoid(opaque_void);
        match secretely_void {}
    }
    x
}
```
Since the `TypeFolder` doesn't recurse into the struct fields, it doesn't reveal the opaque and `uninhabitedness_predicate` thinks it's nonempty.

Reply:
I see three options:
- ignore inconsistency
- recurse through ADT fields too (would require building new `AdtDef`s etc, doesn't sound good)
- make `Ty::inhabited_predicate` work at the function granularity and inspect opaques (would make rustc as a whole more consistent on this front, which may or may not break things)

I'm partial to option 1, especially since the difference cannot be observed without `exhaustive_patterns`. This could be a blocker for its stabilization however.

Reply:
Let's stay with option 1. I would really appreciate if you note this inconsistency in the `exhaustive_patterns` tracking issue, though, for future stabilization to revisit.

Reply:
Will do!

Reply:
(if/when this gets merged)